### PR TITLE
[WIP] Example with multi-GPU for both feature engineering and training

### PR DESCRIPTION
**Pull request by karlhigley**
This is an early draft of a Tensorflow example notebook that combines the existing multi-GPU toy dataset feature engineering example with the recent multi-GPU training examples with Horovod.

### CODE REVIEW

**Comment by karlhigley** [comment]
Not quite sure where I've gone wrong, but I'm currently seeing errors like this after writing out the training script to disk and starting it with `horovodrun`:
```
[1,0]<stderr>:Traceback (most recent call last):
[1,0]<stderr>:  File "trainer.py", line 92, in <module>
[1,0]<stderr>:    tf.feature_column.categorical_column_with_identity(
[1,0]<stderr>:  File "/opt/conda/lib/python3.8/site-packages/tensorflow/python/feature_column/feature_column_v2.py", line 1569, in categorical_column_with_identity
[1,0]<stderr>:    raise ValueError(
[1,0]<stderr>:ValueError: num_buckets 0 < 1, column_name movieId
```

**Comment by karlhigley** [comment]
@bschifferer @rnyak This PR is now in a runnable state with the Tensorflow training container. I've left a few brief notes about what's different here from the existing MovieLens example, but it will probably need a bit more explanation on my part.

File: examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb [hunk 1]
```diff
@@ -0,0 +1,897 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "traditional-kingdom",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!horovodrun -np 3 sh hvd_wrapper.sh python trainer.py --dir_in $BASE_DIR --b_size 1024"
```
Comment (jperez999) [inline examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb:872]: Going to need to change this now to --batch_size

**Review by jperez999 — CHANGES_REQUESTED** [review]

File: examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb (hunk 1 above)
Comment (karlhigley) [reply examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb:872]: I would definitely have forgotten that. Thanks!

**Review by jperez999 — APPROVED** [review]

**Comment by bschifferer** [comment]
I got the error that horovad is not installed in the tensorflow container.

```
(rapids) root@bc00acc2b333:/workspace/01_NVT/30_MultiGPU/NVTabular/examples/multi-gpu-movielens# horovodrun -np 2 sh hvd_wrapper.sh python trainer.py --dir_in basedir --batch_size 1024
2021-04-06 08:34:10.686704: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
[1,1]<stderr>:Traceback (most recent call last):
[1,1]<stderr>:  File "trainer.py", line 13, in <module>
[1,1]<stderr>:    import horovod.tensorflow as hvd  # noqa: E402
[1,1]<stderr>:ModuleNotFoundError: No module named 'horovod'
[1,0]<stderr>:Traceback (most recent call last):
[1,0]<stderr>:  File "trainer.py", line 13, in <module>
[1,0]<stderr>:    import horovod.tensorflow as hvd  # noqa: E402
[1,0]<stderr>:ModuleNotFoundError: No module named 'horovod'
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[26566,1],1]
  Exit code:    1
  ```

**Comment by bschifferer** [comment]
@karlhigley I restructured the notebook to follow the getting-started flow:
1. Download and Convert
2. ETL with NVTabular
3. Training with TensorFlow

I added text for previous notebooks + learnings. I added content to examples/README.md . 

**ToDo**
Can you describe more the new content of the last section "Training with TensorFlow"?
What is "OpenMPI (or Slurm)"?
What is "Horovod"?
How do we use OpenMI/Horovod?
What is Horovod and how does it sync the parameter?
Why do we use seed_fn ?

Does not need to be long paragraphs - just a few sentences and then references to learn more?

**Comment by karlhigley** [comment]
@bschifferer 

> I got the error that horovad is not installed in the tensorflow container.
> I need to install `pip install horovod[tensorflow]`

Ah, forgot to share a step: in order to use the Horovod bindings in the container, you have to `export PYTHONPATH=$PYTHONPATH:/usr/local/lib/python3.8/dist-packages`. I'm not sure if we can do that from inside the notebook, so maybe it would be better to `pip install horovod` in a cell at the top?

**Review by bschifferer — COMMENTED** [review]
> @bschifferer
> 
> > I got the error that horovad is not installed in the tensorflow container.
> > I need to install `pip install horovod[tensorflow]`
> 
> Ah, forgot to share a step: in order to use the Horovod bindings in the container, you have to `export PYTHONPATH=$PYTHONPATH:/usr/local/lib/python3.8/dist-packages`. I'm not sure if we can do that from inside the notebook, so maybe it would be better to `pip install horovod` in a cell at the top?

In many other examples, we haven't provided installation guide, yet (e.g. jupyter notebook, pydot, wget, etc.).
We may provide docker-compose.yml (see criteo example) or we install pip from jupyter notebook or just make a comment in the Training with TensorFlow section

File: examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb [hunk 2]
```diff
@@ -0,0 +1,897 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "traditional-kingdom",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!horovodrun -np 3 sh hvd_wrapper.sh python trainer.py --dir_in $BASE_DIR --batch_size 1024"
```
[the remark this answers is not in the document]
Comment (karlhigley) [reply examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb]: It's currently called `batch_size` in the example scripts: https://github.com/NVIDIA/NVTabular/blob/dd1c8b64cd372e5f2f80b7f79b8ca66410958abc/examples/horovod/tf_hvd_simple.py#L21

**Comment by jperez999** [comment]
We will be fixing the training containers for release, so that horovod is no longer hidden and available on the tf and torch containers.

**Comment by karlhigley** [comment]
@bschifferer I added more description of the parts that are specific to this notebook (e.g. training with Horovod.) Give it a look?

**Comment by rnyak** [comment]
@karlhigley few comments from my side:

1. Add links for OpenMPI and Slurm
2.   getting `<stderr>:trainer.py: error: unrecognized arguments: --b_size 1024`
3.  still getting errror even if I used `--batch_size` flag instead of `--b_size` : `!horovodrun -np 2 sh hvd_wrapper.sh python trainer.py --dir_in $BASE_DIR --batch_size 1024`

```
[1,1]<stderr>:  File "trainer.py", line 31, in <module>
[1,1]<stderr>:    BATCH_SIZE = int(args.b_size) or 16384  # Batch Size
[1,1]<stderr>:AttributeError: 'Namespace' object has no attribute 'b_size'
[1,0]<stderr>:Traceback (most recent call last):
[1,0]<stderr>:  File "trainer.py", line 31, in <module>
[1,0]<stderr>:    BATCH_SIZE = int(args.b_size) or 16384  # Batch Size
[1,0]<stderr>:AttributeError: 'Namespace' object has no attribute 'b_size'
```
This because of this line in the `trainer.py`:  `BATCH_SIZE = int(args.b_size) or 16384  # Batch Size` Needs to be updated as `args.batch_size`.

**Comment by rnyak** [comment]
> Read -1, expected 66304, errno = 1

@karlhigley  I searched for it and found: https://github.com/horovod/horovod/issues/503

```
One option to remove those messages is to filter them out. When you run your training command:

mpirun ... python tensorflow_word2vec.py |& grep -v "Read -1"
```
it might help?

**Comment by karlhigley** [comment]
@rnyak Thanks for the comments! Normally it should print out a training loss once an epoch, so does sound like something is not working properly. What environment/machine are you running the notebook on? (Wondering because the script uses ~100% of the GPU compute capacity on my 3 GPU machine, so it does make the machine work pretty hard.)

**Comment by rnyak** [comment]
> @rnyak Thanks for the comments! Normally it should print out a training loss once an epoch, so does sound like something is not working properly. What environment/machine are you running the notebook on? (Wondering because the script uses ~100% of the GPU compute capacity on my 3 GPU machine, so it does make the machine work pretty hard.)

@karlhigley I used merlin-tensorflow-training container on a 4 X 16 GB V100 system. And I install horovod via `pip install horovod[tensorflow]`

**Review by bschifferer — COMMENTED** [review]
> @bschifferer I added more description of the parts that are specific to this notebook (e.g. training with Horovod.) Give it a look?

Awesome - it explains really great how we use multiGPU dataloaders and what are the differences.

File: examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb (hunk 1 above)
Comment (rnyak) [reply examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb]: @jperez999 @karlhigley you also need to change this line in the notebook where it writes the code in a trainer.py:
`BATCH_SIZE = int(args.b_size) or 16384  # Batch Size` It is still `args.b_size`.

**Pull request merged**